### PR TITLE
#14038: Remove global BUFFER_MAP and make the tracking of buffers local to an allocator

### PR DESCRIPTION
**Pull request by tt-aho**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/14038

### Problem description
We had a global BUFFER_MAP that tracked buffers by device/address. This was bugged since it doesn't distinguish between buffer types, as different buffer types can have the same address. The two users of this map are device, which uses it to deallocate all buffers on device close, and ttnn for dumping buffer statistics. Discussion with @tt-asaigal was that for distributed work in the future, the allocation won't be by device, so we will have the allocator track its own allocated buffers, and device can query this.

### What's changed
Make buffer map local to the allocator object.
Update apis related to allocate/deallocate to not need to pass `bottom_up` separately and make them pass around the buffer ptr, instead of extracting fields at the higher level. This simplifies the api and also needed for tracking the buffers in the allocator.
There is a dependency on #14023 to make buffers thread safe, which would enable the change to pass buffer ptrs around.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by aliuTT — APPROVED** [review]

File: tt_metal/impl/allocator/allocator.hpp
Comment (ayerofieiev-tt) [inline tt_metal/impl/allocator/allocator.hpp:138]: I wonder, why `set`?

File: tt_metal/impl/device/device.cpp
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.cpp:3140]:
```suggestion
std::set<Buffer *> &Device::get_allocated_buffers() const {
```

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (ayerofieiev-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:2789]: Can we move allocation from this `if` statement? 
Seems like the only difference is that one allocates `buffer->size()` size and the other `buffer->num_dev_pages() * buffer->page_size()`

Also.. its probably best to move this logic inside the buffer itself

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (ayerofieiev-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:2810]: I love it that bottom_up is removed from the api. Thank you

File: tt_metal/tt_metal.cpp
Comment (ayerofieiev-tt) [inline tt_metal/tt_metal.cpp:870]: I wonder if it makes sense to add `Buffer:get_allocator()` so same pattern `*(buffer->device()->allocator_)` which we use in many places could be simplified.

File: ttnn/cpp/ttnn/reports.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/reports.hpp:85]: 👍

File: ttnn/cpp/ttnn/reports.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/reports.hpp:189]: Can we do something like this instead?
```
uint32_t bank_id = 0;
for (int page_index = 0; page_index < num_pages; page_index++) {
    tt::tt_metal::CoreLocation core;
    uint64_t page_address = 0;

    if (buffer->buffer_layout() == tt::tt_metal::TensorMemoryLayout::INTERLEAVED) {
        page_address = buffer->page_address(bank_id, page_index);
        core = buffer->logical_core_from_bank_id(bank_id);
        bank_id = (bank_id + 1) % num_banks;
    } else {
        const auto &buffer_page_mapping = *buffer->get_buffer_page_mapping();
        auto dev_page_index = buffer_page_mapping.host_page_to_dev_page_mapping_[page_index];
        core = buffer_page_mapping.all_cores_[buffer_page_mapping.dev_page_to_core_mapping_[dev_page_index]];
        bank_id = device->bank_ids_from_logical_core(buffer->buffer_type(), core)[0];
        page_address = buffer->sharded_page_address(bank_id, dev_page_index);
    }

    BufferPageInfo buffer_page_info = {};
    buffer_page_info.device_id = device_id;
    buffer_page_info.address = address;
    buffer_page_info.core_y = core.y;
    buffer_page_info.core_x = core.x;
    buffer_page_info.bank_id = bank_id;
    buffer_page_info.page_index = page_index;
    buffer_page_info.page_address = page_address;
    buffer_page_info.page_size = page_size;
    buffer_page_info.buffer_type = buffer->buffer_type();

    buffer_page_infos.push_back(buffer_page_info);
}
```

**Review by ayerofieiev-tt — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (tt-aho) [reply tt_metal/impl/dispatch/command_queue.cpp:2789]: I left it like this for now as I recently discussed with @TT-BrianLiu about the issues regarding this and how we want to change it. The reason it's like this is because the `buffer->size()` should reflect the on device size, but for sharded buffers it does not and actually reflects the unpadded/host size, and we have dependencies related to sharding infra that seems to expect it being this way for now. I could add a `buffer::device_size()` api if we want to avoid the branching here, but left it like this since I was expecting Brian to clean it up with his planned changes.

File: ttnn/cpp/ttnn/reports.hpp
Comment (tt-aho) [reply ttnn/cpp/ttnn/reports.hpp:189]: Sure I can clean it up. My original change here added a top level loop but didn't touch the logic so all these changes were just indentation.

File: tt_metal/tt_metal.cpp
Comment (tt-aho) [reply tt_metal/tt_metal.cpp:870]: Hmm, normally most code doesn't directly interact with allocator, there should only be two places, the initial allocate and deallocate buffer calls. Seems like we're calling it in some places where we already have a reference to the allocator, I will clean those up so there's only 2 for now and if it's more common we can add an api

File: tt_metal/impl/allocator/allocator.hpp
Comment (tt-aho) [reply tt_metal/impl/allocator/allocator.hpp:138]: Since we need to have good erasure performance when deallocating buffers, I chose set (and set keeps it comparable to the original insertion/erasure complexity we had with map). Open to thoughts on alternative container to use

File: tt_metal/impl/allocator/allocator.cpp
Comment (dmakoviichuk-tt) [inline tt_metal/impl/allocator/allocator.cpp:381]: why allocated buffers is set but not unordered_set

File: tt_metal/impl/allocator/allocator.hpp
Comment (dmakoviichuk-tt) [reply tt_metal/impl/allocator/allocator.hpp:138]: do you mean that unordered_set has worse deallocation perofromance?

**Review by dmakoviichuk-tt — COMMENTED** [review]
Please don't merge it before I review. I might get some interesting ideas.

File: tt_metal/impl/allocator/allocator.hpp
Comment (tt-aho) [reply tt_metal/impl/allocator/allocator.hpp:138]: No, the original comment wasn't clear on what the question/issue was, and I thought it was asking why I chose a set-like container, not between using set/unordered_set. I will change to `unordered_set`

File: tt_metal/impl/allocator/allocator.cpp
Comment (tt-aho) [reply tt_metal/impl/allocator/allocator.cpp:381]: I will change to unordered_set

**Comment by dmakoviichuk-tt** [comment]
@tt-aho  I made a simple test:
https://godbolt.org/z/5d5GsTr5q
feels like destructor in unordered_set is faster than in std::set.

**Comment by tt-aho** [comment]
> @tt-aho I made a simple test: https://godbolt.org/z/5d5GsTr5q feels like destructor in unordered_set is faster than in std::set.

I've already made the change to unordered_set.

**Comment by dmakoviichuk-tt** [comment]
@tt-aho  overall if you think we have some perf issues with destructors I'd like to know more and probably I can help.

File: ttnn/cpp/ttnn/types.hpp
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/types.hpp:129]: What is the GLOBAL_BUFFER_ADDRESS_MAP can we remove it too?

**Comment by tt-aho** [comment]
I don't think there are any issues with destructor performance. There were just some misunderstandings in the original thread.

File: ttnn/cpp/ttnn/types.hpp
Comment (tt-aho) [reply ttnn/cpp/ttnn/types.hpp:129]: Yes, that is being removed in #14023 which this pr has a dependency on, so this pr is currently in draft. The entire ttnn buffer class is being removed so once the other pr is merged there won't be any changes in this pr to this file. Change here is just to make sure my changes compile before rebasing.

File: tt_metal/tt_metal.cpp
Comment (abhullar-tt) [inline tt_metal/tt_metal.cpp:816]: not in the scope of this PR but maybe we should uplift to use DeviceAddr throughout

**Review by abhullar-tt — APPROVED** [review]

File: tt_metal/tt_metal.cpp
Comment (tt-aho) [reply tt_metal/tt_metal.cpp:816]: I was originally going to change the address fn to return DeviceAddr, but it required updates to most of the op files since the address is passed directly into the RTAs, which uses U32 so I would need to manually cast all the use cases so I backed out the change.

I'll file an issue to track this.

**Pull request merged**
